### ISSUE TITLE
Backward invocation inside ref.listen when updating provider ref is listening to

### ISSUE THREAD

**Issue by eli1stark** [rrousselGit/riverpod#1757]
**Describe the bug**
When we listen for `state.valueA` and update `state.valueB` inside the listener we have a backward invocation:
```
flutter: Listen for 1 2022-10-10 14:32:16.491797
flutter: Listen for 2 2022-10-10 14:32:16.494572
flutter: Listen for 3 2022-10-10 14:32:16.494869
flutter: Listen for 4 2022-10-10 14:32:16.495141
flutter: Listen for 5 2022-10-10 14:32:16.495383
flutter: Listen for 5 2022-10-10 14:32:16.495852
flutter: Listen for 4 2022-10-10 14:32:16.496019
flutter: Listen for 3 2022-10-10 14:32:16.496196
flutter: Listen for 2 2022-10-10 14:32:16.496375
flutter: Listen for 1 2022-10-10 14:32:16.496551
```

Expected invocation:
```
flutter: Listen for 1 2022-10-10 14:32:57.941396
flutter: Listen for 1 2022-10-10 14:32:57.944069
flutter: Listen for 2 2022-10-10 14:32:57.944260
flutter: Listen for 2 2022-10-10 14:32:57.944516
flutter: Listen for 3 2022-10-10 14:32:57.944711
flutter: Listen for 3 2022-10-10 14:32:57.944885
flutter: Listen for 4 2022-10-10 14:32:57.945098
flutter: Listen for 4 2022-10-10 14:32:57.945297
flutter: Listen for 5 2022-10-10 14:32:57.945482
flutter: Listen for 5 2022-10-10 14:32:57.945666
```

**To Reproduce**
hooks_riverpod: 2.0.2
freezed: 2.1.0+1

```dart
import 'package:flutter/material.dart';
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:hooks_riverpod/hooks_riverpod.dart';

part 'main.freezed.dart';

void main() {
  runApp(const ProviderScope(child: MyApp()));
}

class MyApp extends ConsumerWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return MaterialApp(
      home: Scaffold(
        body: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: const [
                AnimatedText(index: 1),
                AnimatedText(index: 2),
                AnimatedText(index: 3),
                AnimatedText(index: 4),
                AnimatedText(index: 5),
              ],
            ),
            TextButton(
              child: const Text('Start'),
              onPressed: () {
                ref.read(textNotipod.notifier).plus();
              },
            ),
          ],
        ),
      ),
    );
  }
}

class AnimatedText extends ConsumerWidget {
  const AnimatedText({
    required this.index,
    super.key,
  });

  final int index;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    ref.listen(textNotipod.select((s) => s.plus), (prev, curr) {
      print('Listen for $index ${DateTime.now()}');

      ref.read(textNotipod.notifier).minus();

      print('Listen for $index ${DateTime.now()}');
    });

    return Text(
      index.toString(),
    );
  }
}

@freezed
class TextState with _$TextState {
  const factory TextState({
    required int minus,
    required int plus,
  }) = _TextState;
}

class TextNotifier extends StateNotifier<TextState> {
  TextNotifier() : super(const TextState(plus: 0, minus: 0));

  void plus() => state = state.copyWith(plus: state.plus + 1);
  void minus() => state = state.copyWith(minus: state.minus - 1);
}

final textNotipod = StateNotifierProvider<TextNotifier, TextState>(
  (ref) {
    return TextNotifier();
  },
);
```

**Comment by rrousselGit** [comment]
That seems expected. Listeners are invoked synchronously on change. 

Why should it print what you asked?

**Comment by eli1stark** [comment]
Why it shouldn't? You said yourself that they are synchronous, print is synchronous, and the update to the state is synchronous as well. From my understanding, the whole listener block should be completed before going to the next one. 
We are subscribed for valueA, update valueB. Why the update of the state freezes the whole listener block execution? 

If we change: `void minus() => state = state.copyWith(minus: state.minus - 1);`  to `void minus() {}` it will work as expected. Or let's imagine another situation, if we, for example, will update `textNotipod2` inside  `textNotipod` listener instead of `textNotipod` it will work as expected as well (1,1,2,2..).

Am I missing something?

**Comment by TimWhiting** [comment]
What you are missing is that the ```state = something``` call actually executes more code than just assigning that variable. It notifies any listeners which causes your listener to execute. (Basically a recursive function call). So just like a recursive function call you will not execute your second print statement until the recursion backs up.

**Comment by rrousselGit** [comment]
>  From my understanding, the whole listener block should be completed before going to the next one.

You're notifying listeners within a listener. You did the equivalent of:

```dart
    ref.listen(textNotipod.select((s) => s.plus), (prev, curr) {
      print('Listen for $index ${DateTime.now()}');

      for (final listener in listeners) {
        listener(curr, curr - 1);
      }

      print('Listen for $index ${DateTime.now()}');
    });

```

So the current output is logical.

**Comment by eli1stark** [comment]
This is recursion in your example, in my example, there is no recursion, I have 5 widgets, and each one of them has 1 listener, 

and I have 1 trigger:
```dart
 onPressed: () {
    ref.read(textNotipod.notifier).plus();
  },
 ```
`plus()` will trigger 1 call in each of 5 widgets, 5 calls total.

`minus()` inside a listener shouldn't do anything to the listener itself since the listener is listening for `plus` using `select`, the only thing it will do is update the `minus` in the state, which again won't do a thing since we are not listening for `minus` in this case.

So, I'm not notifying anyone by updating the state. I just update the state of the notifier, the provider will register the update and the listener will ignore this update since the listener is listening for a different field.

**Comment by rrousselGit** [comment]
My bad, you're right, I got confused.
I'll look into it

An immediate theory I have is that it's an issue with `select`. I guess that when calling `minus()`, this flushes the listeners, and `select` has a new chance to trigger its listeners. So selectors that have yet to notify their listeners (because another index is still pending) have another chance to notify their listeners.

That's a funky edge-case.

**Comment by rrousselGit** [comment]
But I don't think it was intended to be able to notify listeners within a listener.
I'm pretty sure I made the conscious decision to disable that

It's possible that a regression enabled this scenario.

I'll need to double check.

**Comment by rrousselGit** [comment]
Thanks for your time and the example btw. These are pretty rare. I should've run it sooner

**Comment by rrousselGit** [comment]
I don't think anything can be done here.

I don't think it's really a problem either. I would discourage modifying providers within listeners.  
Pretty sure this wasn't allowed before but got enabled when I did performance optimization for listener notification.

So I'd lean toward "won't fix"

**Comment by eli1stark** [comment]
Seems to me that it can be a side effect of the bigger problem that we don't know about yet, usually it is always like that.

My problem can be fixed by allocating a separate provider. If we reference my example, then I need to create a separate provider for `minus`, and in the end, I will have 2 notifiers and providers. And in case I will want to use `minus` in the `TextNotifer` I just need to access it with `StateNotifierProviderRef` like `ref.read(minusPod)`.

This means that the problem here is not about modifying providers inside the listener in general, this problem is about modifying specifically the provider I'm listening to which is a pretty common scenario after you introduced `listenSelf`, but I guess `listenSelf` is different in this case since this happens inside the body of provider, but still worth to check it out just in case. 

Mentally as a riverpod user, having something like `listenSelf` passively saying to me that it's ok to do things like that. It's more a tip for current docs, maybe to emphasize this issue, like "hey, this is a little bit different". 

When you are saying "I would discourage modifying providers within listeners" do you mean any provider or do you mean only the provider we are listening to? In any case, if we lean toward "won't fix" we need to have an `Exception` thrown when we try to update the provider inside the listener, either for all providers or just the one we listen to, it will depend on your answer. I would just throw for the one we listen to since there is no problem with updating other providers.

Without `Exception`, it's only a matter of time before another issue will be created. And by the way, this bug can be hard to spot in the app. And it can break your logic, when you need to do something sequentially, like animation in several different widgets that are invoked by the listener in a specific order.

**Comment by rrousselGit** [comment]
Modifying any provider within any listener (including things like onDispose) is discouraged.    
We could throw, yes. But that would be silently breaking some apps, and probably wouldn't catch all cases.

I don't think it's such a big deal that it's worth releasing a 3.0 just for this.

**Comment by eli1stark** [comment]
Can we mark it `3.0` then, and when `3.0` comes this will be added?
